### ISSUE TITLE
B-JointSP evaluation

### ISSUE THREAD

**Issue by stefanbschneider** [RealVNF/bjointsp-adapter#14]
Discussion on how to best evaluate B-JointSP and compare it against other algorithms (mostly DDPG). Related to https://github.com/RealVNF/evaluation/issues/29

Current approach:

* Fix run duration to 6 (instead of 100 like DDPG) and pass info from all incoming flows within the 6 timesteps as input to B-JointSP
* Reason: B-JointSP treats all specified flows as long-running parallel flows competing for resources. 
* But flows arriving more than 6 time steps after another are sequential and do not compete for resources. If we gave all these flows as input B-JointSP would over-estimate the resource requirements, conclude that there are not enough resources, and place all VNFs uniformly to minimize-over subscription
* Resulting placements and mappings of flows are then converted to placments and scheduling rules that are passed to the simulator
* Another issue with runs of length 6:
  * If no flows arrive during a run, B-JointSP is not even called, but the run is simply skipped without changing the schedule
  * But with random arrival if flows arrive for one ingress but not for another,  B-JointSP removes all VNFs for the unused ingress. The next following flows will then be dropped because no VNFs are available for the ingress.

Alternative:
* Goal: Use same run duration for both DDPG and B-JointSP. DDPG doesn't work well with run duration 6.(?) So let's set run duration 100 for both
* Calculate the average data rate during the run in the simulator and provide it to the B-JointSP adapter. Eg, 10 flows of size 1 within 100 time steps (det. arrival) --> Rate of 10/100 = 0.1
* Take the processing time (VNF delay + flow length) into account, which affects the required resources (in our case 6). --> `0.1 * 6 = 0.6`. Round up to 1.
* Divide by flow size (1) to calculate the number of flows. Here: `1/1=1` (but may be more during bursts or MMP).
* Thus pass 1 flow of size 1 to the B-JointSP

This way, we would calculate roughly the data rate that needs to be served in parallel. But still split it into separate flows of the pre-defined size (here, 1) such that B-JointSP can embed the flows separately.

Does that make sense? @adnan904 @qarawlus

**Comment by stefanbschneider** [comment]
If we implement this, keep the current calculation of flows (the current `master`) of the adapter in a separate branch such that we can switch back quickly and easily.

**Comment by stefanbschneider** [comment]
The description above is pretty unclear and not intuitive. While the calculation works, here's a better description of what we're doing:

* B-JointSP expects flows that are specified as input to run in parallel and be competing for resources
* So, we're trying to calculate the number of overlapping flows within a given `run_duration` that overlap
* Rather than taking the maximum or something more sophisticated, we're calculating the avg. number of overlapping/parallel flows competing for resources
* A flow is competing for resources while it's being processed by a VNF
* Processing a flow takes `flow_length + vnf_processing` time steps. In our case, this is always 1 + 5 = 6
* If we have a run duration of 100, during which 50 flows arrive, then in total the 50 flows are processed for 50x6 = 300 time steps. Since the run duration is only 100, in average 300/100 = 3 flows are being processed in parallel, competing for resources. Thus, we should specify 3 flows, each with size 1, as input to B-Jointsp.
* In general, individually for each ingress, compute `avg_num_parallel_flows = (num_flows * flow_processing_length) / run_duration`. Then create `avg_num_parallel_flows` many flows, each with the specified flow size (in our case 1).

**Comment by stefanbschneider** [comment]
@adnan904 If you implemented and tested this successfully, could you please merge it into the main bjtiontsp-adapter repo (separate branch `avg_flows_over_run`)? I think it's currently just in your fork.

Then we can close this issue
